### DISCUSSION TITLE
Why do extra heap allocations happen?

### DISCUSSION THREAD

**Discussion by CADBIMDeveloper** [dotnet/runtime#90073] [General]
Hi Experts!

I'm playing with the following code, benchmarking heap allocations and comparing performance of heap vs stack allocations. `Test` method has no meaning except it accesses the array, reading and writing elements from it:

```cs
private const int Size = 5000;
private const int Cycles = 1000;

[Benchmark]
public void TestHeapArray()
{
	Span<int> array = new int[Size];

	Test(array);
}

[Benchmark]
public void TestStackAllocArray()
{
	Span<int> array = stackalloc int[Size];

	Test(array);
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static int Test(Span<int> array)
{
	for (var i = 0; i < array.Length; ++i)
		array[i] = i % 10;

	for (var i = 0; i < Cycles; ++i)
	{
		for (var j = 1; j < array.Length - 1; ++j) 
			array[j] = array[j] - array[j - 1] + array[j + 1];
	}

	var sum = 0;

	foreach (var elem in array) 
		sum += elem;

	return sum;
}
```

When I'm changing the Cycles const I see the following results:
Cycles = 10:
```
|              Method |     Mean |   Error |  StdDev |   Gen0 | Allocated |
|-------------------- |---------:|--------:|--------:|-------:|----------:|
|       TestHeapArray | 161.3 us | 0.99 us | 0.88 us | 6.3477 |   20024 B |
| TestStackAllocArray | 157.2 us | 1.31 us | 1.09 us |      - |         - |
```

Cycles = 100:
```
|              Method |     Mean |     Error |    StdDev |   Gen0 | Allocated |
|-------------------- |---------:|----------:|----------:|-------:|----------:|
|       TestHeapArray | 1.100 ms | 0.0217 ms | 0.0266 ms | 5.8594 |   20025 B |
| TestStackAllocArray | 1.107 ms | 0.0216 ms | 0.0317 ms |      - |       1 B |
```
Cycles = 1000:
```
|              Method |     Mean |    Error |   StdDev | Allocated |
|-------------------- |---------:|---------:|---------:|----------:|
|       TestHeapArray | 10.17 ms | 0.094 ms | 0.088 ms |   20033 B |
| TestStackAllocArray | 10.13 ms | 0.097 ms | 0.091 ms |       9 B |
```

So my question is: why does CLR allocate extra few bytes on the heap?

**Comment by EgorBo** [comment]
Reproduces even with TieredCompilatipn=0, codegen for both `TestStackAllocArray` and `Test` doesn't contain any allocations so I don't know - @adamsitnik @timcassell any idea?

**Reply by timcassell** [reply]
We aren't allocating anything in the engine. It's probably somewhere in the runtime. It used to be a much bigger issue pre-net7.0. See https://github.com/dotnet/BenchmarkDotNet/pull/1543 for more information.

**Comment by adamsitnik** [comment]
@timcassell is right, it's not BDN fault as the Engine itself does not allocate any managed memory (https://github.com/dotnet/BenchmarkDotNet/issues/1542)

You need to use a memory profiler to find out. VS Memory Profiler disables Tiered JIT by default, so you need to use PerfView or some other profiler.

For PerfView, you need to use the .NET Alloc CheckBox:

```
The .NET Alloc CheckBox - Causesan event to be fired every time a .NET object is allocated.This can be also activated by the /DotNetAlloc command line option.Note that this only affect processes that start AFTER data collection has started.See GC Heap Net Mem for more. 
This option tends to have a VERY noticeable impact on performance (2X or more). Also,if the application allocates aggressively, so many events will be fired so quickly thatevents will be lost even when the/BufferSizeMB qualifier is used to set the size very large (e.g. 500Meg). 
```

![image](https://github.com/dotnet/runtime/assets/6011991/c4bd0d8f-58f5-4000-98a2-e5d085104d40)

If you can not find any allocations, some old bug like https://github.com/dotnet/runtime/issues/7638 can be back
